### PR TITLE
Implement `no_blank_line_before_class_docstring` preview style

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR implements the `no_blank_line_before_class_docstring` preview style.

## Test Plan

Update existing snapshots.

### Formatter ecosystem

`main`

| project        | similarity index  | total files       | changed files     |
|----------------|------------------:|------------------:|------------------:|
| cpython        |           0.75804 |              1799 |              1648 |
| django         |           0.99984 |              2772 |                34 |
| home-assistant |           0.99955 |             10596 |               213 |
| poetry         |           0.99905 |               321 |                15 |
| transformers   |           0.99967 |              2657 |               324 |
| twine          |           1.00000 |                33 |                 0 |
| typeshed       |           0.99980 |              3669 |                18 |
| warehouse      |           0.99976 |               654 |                14 |
| zulip          |           0.99958 |              1459 |                36 |

`dhruv/no-blank-line-docstring`

| project        | similarity index  | total files       | changed files     |
|----------------|------------------:|------------------:|------------------:|
| cpython        |           0.75804 |              1799 |              1648 |
| django         |           0.99984 |              2772 |                34 |
| home-assistant |           0.99955 |             10596 |               213 |
| poetry         |           0.99905 |               321 |                15 |
| transformers   |           0.99967 |              2657 |               324 |
| twine          |           1.00000 |                33 |                 0 |
| typeshed       |           0.99980 |              3669 |                18 |
| warehouse      |           0.99976 |               654 |                14 |
| zulip          |           0.99958 |              1459 |                36 |

fixes: #8888

### CODE REVIEW

**Comment by T-256** [comment]
It seems ecosystem reported numbers are wrong.
Also, links pointed lines aren't exact.

File: crates/ruff_python_formatter/tests/snapshots/format@statement__class_definition.py.snap [hunk 1]
```diff
@@ -513,7 +513,30 @@ class QuerySet(AltersData):
...
+ 
+ class Test:
+-
+     """Docstring"""
+ 
+     x = 1
+ 
+ 
+ class Test:
+-
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/tests/snapshots/format@statement__class_definition.py.snap:535]: Can we add a few tests where we have comments before the docstring (including new or empty lines)

**Review by MichaReiser — COMMENTED** [review]
Could you update your pr summary and include the 'old' compatibility numbers. I don't expect them to change but just to be safe

**Comment by charliermarsh** [comment]
The ecosystem line numbers are not guaranteed to be exact (hence the tilde). It’s a diff of two changes against HEAD, so mapping that diff back to a line number in HEAD is not trivial. (Not sure about the “number of lines changed” though.)

**Review by konstin — APPROVED** [review]

File: crates/ruff_python_formatter/tests/snapshots/format@statement__class_definition.py.snap (hunk 1 above)
Comment (dhruvmanila) [reply crates/ruff_python_formatter/tests/snapshots/format@statement__class_definition.py.snap:535]: I've tested it out locally with different combinations of comments, newline and docstrings. It seems that the Black preview style is a bit inconsistent with the stable style.

<details><summary>Original source code:</summary>
<p>

```python
class OneBlankLineDocstring:

    """This is a docstring."""


class DocstringWithComment0:
    # This is a comment
    """This is a docstring."""


class DocstringWithComment1:
    # This is a comment

    """This is a docstring."""


class DocstringWithComment2:

    # This is a comment
    """This is a docstring."""


class DocstringWithComment3:

    # This is a comment

    """This is a docstring."""


class DocstringWithComment4:


    # This is a comment


    """This is a docstring."""
```

</p>
</details> 

<details><summary>Black stable formatting:</summary>
<p>

```python
class NormalDocstring:

    """This is a docstring."""


class DocstringWithComment0:
    # This is a comment
    """This is a docstring."""


class DocstringWithComment1:
    # This is a comment

    """This is a docstring."""


class DocstringWithComment2:
    # This is a comment
    """This is a docstring."""


class DocstringWithComment3:
    # This is a comment

    """This is a docstring."""


class DocstringWithComment4:
    # This is a comment

    """This is a docstring."""
```

</p>
</details> 

<details><summary>Black preview formatting:</summary>
<p>

```python
class OneBlankLineDocstring:
    """This is a docstring."""


class DocstringWithComment0:
    # This is a comment
    """This is a docstring."""


class DocstringWithComment1:
    # This is a comment

    """This is a docstring."""


class DocstringWithComment2:

    # This is a comment
    """This is a docstring."""


class DocstringWithComment3:

    # This is a comment

    """This is a docstring."""


class DocstringWithComment4:

    # This is a comment

    """This is a docstring."""
```

</p>
</details>

What I'm noticing is that the rules change if there's a comment in between the class header and docstring in a way that for `n1` and `n2` number of blank lines before a comment and docstring respectively the following holds true:
1. If `n1` / `n2` is 0, keep it 0
2. If `n1` / `n2` is 1, keep it 1
3. If `n1` / `n2` > 1, then reduce it to 1

It's probably another preview style which is keeping the newline before the comment (`allow_empty_first_line_before_block_or_comment`).

**Review by MichaReiser — APPROVED** [review]

**Pull request merged**